### PR TITLE
Kamu 99: Eliminate improper type casts in lineage graph component

### PR DESCRIPTION
**Pull request by zaychenko-sergei**
- Reworked datastructures used to prepare lineage graph data
- Unified naming of lineage-related items
- Restructured lineage graph update responsibilities in similar manner as other dataset tabs
- Eliminated redundand lineage graph resets

### CODE REVIEW

File: src/app/dataset-view/dataset.service.ts [hunk 1]
```diff
@@ -303,66 +230,140 @@ export class AppDatasetService {
...
+                                            return {
+                                                basics: downDependency3 as DatasetBasicsFragment,
+                                                downstreamDependencies: downDependency3.metadata.currentDownstreamDependencies.map(
+                                                    (downDependency4) => {
+                                                        return {
+                                                            basics: downDependency4 as DatasetBasicsFragment,
+                                                            downstreamDependencies: downDependency4.metadata.currentDownstreamDependencies.map(
+                                                                (downDependency5) => {
+                                                                    return {
+                                                                        basics: downDependency5 as DatasetBasicsFragment,
```
Comment (sergiimk) [inline src/app/dataset-view/dataset.service.ts:266]: This probably could be expressed more concisely through recursion

File: src/app/dataset-view/dataset.service.ts [hunk 2]
```diff
@@ -303,66 +230,140 @@ export class AppDatasetService {
...
+                            }
+                        )
+                    };
+                }
+            )
+        };
+     }
+
+    private updatelineageGraph(origin: DatasetLineageNode) {
+        const lineageGraphEdges: DatasetBasicsFragment[][] = [];
```
Comment (sergiimk) [inline src/app/dataset-view/dataset.service.ts:333]: Should the dealings with nodes and edges be a concern of Lineage component, not dataset service? `origin` node seems to contain everything that is needed to vizualize the graph and the way it happens is not a service-level concern?

**Review by sergiimk — COMMENTED** [review]
Great improvement :+1:

File: src/app/dataset-view/dataset.service.ts (hunk 1 above)
Comment (zaychenko-sergei) [reply src/app/dataset-view/dataset.service.ts:266]: The whole point of this is that there is no way to do it in a completely type-safe manner. First of all, there are no explicit types of parts, those generated union/intersection types do not have any name. Second each level is a different data type. So, we cannot really express it recursively with how the base GraphQL query is defined.

File: src/app/dataset-view/dataset.service.ts (hunk 2 above)
Comment (zaychenko-sergei) [reply src/app/dataset-view/dataset.service.ts:333]: At least building a traditional graph data structure should be done in the service on my opinion. The service should hide the complexity of not completely adequate GraphQL structure. But the bondary can be moved as we revise it again.

**Pull request merged**